File: packages/core/src/capabilities.ts
/** Feature flags reported by a yugendb driver. */
export interface Capabilities {
  /** Driver supports transactional operation groups. */
  readonly transactions: boolean;
  /** Driver supports time-to-live expiry. */
  readonly ttl: boolean;
  /** Driver supports prefix scans. */
  readonly prefixScan: boolean;
  /** Driver supports atomic increment operations. */
  readonly atomicIncrement: boolean;
  /** Driver supports batch writes. */
  readonly batchWrite: boolean;
  /** Driver exposes a raw SQL escape hatch. */
  readonly rawSql: boolean;
  /** Driver exposes a document query escape hatch. */
  readonly documentQuery: boolean;
  /** Driver exposes JSON query behaviour. */
  readonly jsonQuery: boolean;
  /** Driver supports migration helpers. */
  readonly migrations: boolean;
  /** Driver manages connection pooling. */
  readonly connectionPooling: boolean;
  /** Driver supports watch or subscription behaviour. */
  readonly watch: boolean;
  /** Driver supports backup helpers. */
  readonly backup: boolean;
}

/** Returns the minimal capability set required of every driver. */
export function minimalCapabilities(): Capabilities {
  return {
    transactions: false,
    ttl: false,
    prefixScan: false,
    atomicIncrement: false,
    batchWrite: false,
    rawSql: false,
    documentQuery: false,
    jsonQuery: false,
    migrations: false,
    connectionPooling: false,
    watch: false,
    backup: false,
  };
}

/** Returns the capability set for the in-memory driver. */
export function memoryCapabilities(): Capabilities {
  return {
    ...minimalCapabilities(),
    ttl: true,
    prefixScan: true,
    batchWrite: true,
  };
}

/** Returns the capability set for the SQLite driver. */
export function sqliteCapabilities(): Capabilities {
  return {
    ...minimalCapabilities(),
    transactions: true,
    ttl: true,
    prefixScan: true,
    batchWrite: true,
    migrations: true,
  };
}

File: crates/yugendb-core/src/capabilities.rs
//! Driver capability reporting.

/// Feature report exposed by each driver.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Capabilities {
    /// Driver supports transactional operation groups.
    pub transactions: bool,
    /// Driver supports time-to-live expiry.
    pub ttl: bool,
    /// Driver supports prefix scans.
    pub prefix_scan: bool,
    /// Driver supports atomic increment operations.
    pub atomic_increment: bool,
    /// Driver supports batch writes.
    pub batch_write: bool,
    /// Driver exposes a raw SQL escape hatch.
    pub raw_sql: bool,
    /// Driver exposes a document query escape hatch.
    pub document_query: bool,
    /// Driver exposes JSON query behaviour.
    pub json_query: bool,
    /// Driver supports migration helpers.
    pub migrations: bool,
    /// Driver manages connection pooling.
    pub connection_pooling: bool,
    /// Driver supports watch or subscription behaviour.
    pub watch: bool,
    /// Driver supports backup helpers.
    pub backup: bool,
}

impl Capabilities {
    /// Minimal capability set required of every driver.
    #[must_use]
    pub const fn minimal() -> Self {
        Self {
            transactions: false,
            ttl: false,
            prefix_scan: false,
            atomic_increment: false,
            batch_write: false,
            raw_sql: false,
            document_query: false,
            json_query: false,
            migrations: false,
            connection_pooling: false,
            watch: false,
            backup: false,
        }
    }

    /// Capability shape for the memory driver.
    #[must_use]
    pub const fn memory() -> Self {
        Self {
            transactions: false,
            ttl: true,
            prefix_scan: true,
            atomic_increment: false,
            batch_write: true,
            raw_sql: false,
            document_query: false,
            json_query: false,
            migrations: false,
            connection_pooling: false,
            watch: false,
            backup: false,
        }
    }

    /// Capability shape for the SQLite driver.
    #[must_use]
    pub const fn sqlite() -> Self {
        Self {
            transactions: true,
            ttl: true,
            prefix_scan: true,
            atomic_increment: false,
            batch_write: true,
            raw_sql: false,
            document_query: false,
            json_query: false,
            migrations: true,
            connection_pooling: false,
            watch: false,
            backup: false,
        }
    }
}

impl Default for Capabilities {
    fn default() -> Self {
        Self::minimal()
    }
}
